quantity/core.py

```diff
@@ -0,0 +1,243 @@
+
+        other_value, other_unit = get_value_and_unit(other)
+        self_value = self.value
+        (conv0, conv1), unit = units_helper(self.unit, other_unit)
+        if conv0 is not None:
+            self_value = conv0(self_value)
+        if conv1 is not None:
+            other_value = conv1(other_value)
+        try:
+            value = getattr(self_value, op)(other_value)
+        except AttributeError:
+            return NotImplemented
+
+        return replace(self, value=value, unit=unit)
+
+    __array_ufunc__ = None
+    __array_function__ = None
+
+
+DEFERRED = "dtype", "device", "ndim", "shape", "size"
+SAME_UNIT = "mT", "T", "__abs__", "__neg__", "__pos__", "__getitem__", "to_device"
+OPERATORS = (
+    "__add__",
+    "__floordiv__",
+    "__matmul__",
+    "__mod__",
+    "__mul__",
+    "__pow__",
+    "__sub__",
+    "__truediv__",
+)
+COMPARISONS = "__eq__", "__ge__", "__gt__", "__le__", "__lt__", "__ne__"
+DEFER_DIMENSIONLESS = "__complex__", "__float__", "__int__"
+NOT_IMPLEMENTED = (
+    "__and__",
+    "__bool__",
+    "__index__",
+    "__invert__",
+    "__lshift__",
+    "__or__",
+    "__rshift__",
+    "__xor__",
+)
+TODO = "__dlpack__", "__dlpack_device__"
+
+
+def make_deferred(attr):
+    # Use array_api_compat getter if available (size, device), since
+    # some array formats provide inconsistent implementations.
+    attr_getter = getattr(array_api_compat, attr, operator.attrgetter(attr))
+    def deferred(self):
+        return attr_getter(self.value)
+
+    return deferred
+
+
+for attr in DEFERRED:
+    setattr(Quantity, attr, property(make_deferred(attr)))
+
+
+def make_same_unit_method(attr):
+    if array_api_func := getattr(array_api_compat, attr, None):
+        def same_unit(self, *args, **kwargs):
+            return replace(self, value=array_api_func(self.value, *args, **kwargs))
+
+    else:
+        def same_unit(self, *args, **kwargs):
+            return replace(self, value=getattr(self.value, attr)(*args, **kwargs))
+
+    return same_unit
+
+
+def make_same_unit_attribute(attr):
+    attr_getter = getattr(array_api_compat, attr, operator.attrgetter(attr))
+    def same_unit(self):
+        return replace(self, value=attr_getter(self.value))
+
+    return property(same_unit)
+
+
+for attr in SAME_UNIT:
+    if attr in {"mT", "T"}:
+        same_unit = make_same_unit_attribute(attr)
+    else:
+        same_unit = make_same_unit_method(attr)
+    setattr(Quantity, attr, same_unit)
+
+
+def make__op__(op, helper):
+    if op.startswith("__r"):
+        op_func = getattr(operator, op.replace("__r", "__"))
+
+        def wrapped_helper(u1, u2):
+            return helper(op_func, u2, u1)
+    else:
+        op_func = getattr(operator, op)
+
+        def wrapped_helper(u1, u2):
+            return helper(op_func, u1, u2)
+
+    def __op__(self, other):
+        return self._operate(other, op, wrapped_helper)
+
+    return __op__
+
+
+for attr in OPERATORS:
```

Comment:
This is good for testing, but it's too dynamic to be understood statically. These can also be easily set directly on Q2, if the `make__op__` is well typed.

```python
class Quantity:
    # 2 arg ops
    __add__ = make_2argop_("__add__")
    __radd__ = make_2argop_("__radd__")
    __iadd__ = make_2argop_("__iadd__")
   
```

which isn't overly burdensome to enumerate on the class.

Reply:
I've gone ahead and now set all things directly on the class. However, I don't quite understand how I would need to type this, so maybe that can be for follow-up? By you?? :heart:

Reply:
> By you?? 

Sure!